subsys/bluetooth/mesh/sensor_types.c

```diff
@@ -0,0 +1,952 @@
+	{                                                                      \
+		.flags = ((_flags) | (SCALAR_IS_DIV(_scalar) ? DIVIDE : 0)),   \
+		.max = _max,                                                   \
+		.value = (s64_t)((SCALAR_IS_DIV(_scalar) ? (1.0 / (_scalar)) : \
+							   (_scalar)) +        \
+				 0.5),                                         \
+	}
+
+#define SCALAR_REPR(_scalar, _flags) SCALAR_REPR_RANGED(_scalar, _flags, 0)
+
+#ifdef CONFIG_BT_MESH_SENSOR_LABELS
+
+#define SCALAR_FORMAT_MAX(_size, _flags, _unit, _scalar, _max)                 \
+	{                                                                      \
+		.unit = &bt_mesh_sensor_unit_##_unit,                          \
+		.encode = scalar_encode, .decode = scalar_decode,              \
+		.size = _size,                                                 \
+		.user_data = (void *)&(                                        \
+			(const struct scalar_repr)SCALAR_REPR_RANGED(          \
+				_scalar, ((_flags) | HAS_MAX), _max)),         \
+	}
+
+#define SCALAR_FORMAT(_size, _flags, _unit, _scalar)                           \
+	{                                                                      \
+		.unit = &bt_mesh_sensor_unit_##_unit,                          \
+		.encode = scalar_encode, .decode = scalar_decode,              \
+		.size = _size,                                                 \
+		.user_data = (void *)&((const struct scalar_repr)SCALAR_REPR(  \
+			_scalar, _flags)),                                     \
+	}
+#else
+
+#define SCALAR_FORMAT_MAX(_size, _flags, _unit, _scalar, _max)                 \
+	{                                                                      \
+		.encode = scalar_encode, .decode = scalar_decode,              \
+		.size = _size,                                                 \
+		.user_data = (void *)&(                                        \
+			(const struct scalar_repr)SCALAR_REPR_RANGED(          \
+				_scalar, ((_flags) | HAS_MAX), _max)),         \
+	}
+
+#define SCALAR_FORMAT(_size, _flags, _unit, _scalar)                           \
+	{                                                                      \
+		.encode = scalar_encode, .decode = scalar_decode,              \
+		.size = _size,                                                 \
+		.user_data = (void *)&((const struct scalar_repr)SCALAR_REPR(  \
+			_scalar, _flags)),                                     \
+	}
+#endif
+
+enum scalar_repr_flags {
+	UNSIGNED = 0,
+	SIGNED = BIT(1),
+	DIVIDE = BIT(2),
+	/** The highest encoded value represents "undefined" */
+	HAS_UNDEFINED = BIT(3),
+	/**
+	 * The second highest encoded value represents
+	 * "value is higher than xxx".
+	 */
+	HAS_HIGHER_THAN = (BIT(4)),
+	/** The second highest encoded value represents "value is invalid" */
+	HAS_INVALID = (BIT(5)),
+	HAS_MAX = BIT(6),
+};
+
+struct scalar_repr {
+	enum scalar_repr_flags flags;
```

Comment:
Is the scalar definition and utilities only applicable to sensor?
Just wondering if this is something that is re-used by other models and should be moved to mesh model types.

Reply:
Good point. It's actually applicable to the properties models as well, but I didn't want to break that module in this PR. Made a task for this to follow up later.